GraphQL: Dynamic block support (#8598)
magento/devdocs: src/guides/v2.4/graphql/queries/dynamic-blocks.md

---
group: graphql
title: dynamicBlocks query
---

> Dynamic blocks can be inserted in CMS pages and blocks using widgets. The `dynamicBlocks` query returns the content of dynamic blocks that implement the Dynamic Blocks Rotator widget.

Reviewer: we have to make it more obvious that Dynamic Blocks have to be defined in the Dynamic Blocks section of the admin and Dynamic Blocks Rotator is a widget inserted in a CMS page.
Purpose of the new gql query that we added is to grab content of what the widget is returning by parsing some html and extract those parameters that you describe further

That means doing some reference of how the Dynamic Blocks Rotator widget looks like in html when is returned by the cmsPage query.

You can also define widgets

Reviewer 2: I would promote linking to the user guides for how to setup Dynamic Blocks in this section to help provide context. Also note we renamed "Banner" to "Dynamic Block" in 2.3.1 or 2.3.0, so may need to briefly mention that too.

---
group: graphql
title: dynamicBlocks query
---

The `dynamicBlocks` query returns a list of dynamic blocks that have been placed in a Dynamic Blocks Rotator inline widget and meet the specified criteria.

{:.bs-callout-info}
The Banner functionality was removed from Magento 2.4.0 and replaced with dynamic blocks.

> When a [Dynamic Blocks Rotator inline widget is created](https://docs.magento.com/user-guide/cms/dynamic-blocks.html), the Magento administrator can select the following options:

Reviewer: link to https://docs.magento.com/user-guide/cms/dynamic-blocks-rotate.html instead

*  **Specified Dynamic Blocks**
*  **Cart Price Rule Related**
*  **Catalog Price Rule Related**

> Widgets defined with the **Specified Dynamic Blocks** option affect CMS page rendering. The other two options are used for rendering cart, product, and catalog pages and are not applicable for PWA applications. Therefore, in most cases, your query should assign the value of `SPECIFIED` to the `type` input attribute.

Reviewer: in most cases for a CMS page

you also should add that Cart Price Rule Related
Catalog Price Rule Related don't really have direct use cases for CMS page

{:.bs-callout-info}
Magento GraphQL supports the **Display all instead of rotating** rotation mode only.

## Syntax

```graphql
dynamic_blocks(
    input: DynamicBlocksFilterInput
    pageSize: Int
    currentPage: Int
    ): DynamicBlocks!
```

## Example usage

> The following query returns all dynamic blocks of type `SPECIFIED`. The returned item is a a dynamic block containing only text. The second item contains a PNG file.

Reviewer 3: ```suggestion
The following query returns all dynamic blocks of type `SPECIFIED`. The returned item is a dynamic block containing only text. The second item contains a PNG file.
```

**Request:**

```graphql
{
  dynamicBlocks(input: 
  {
    type: SPECIFIED
  })
  {
    items {
      uid
      content {
        html
      }
    }
    page_info {
      current_page
      page_size
      total_pages
    }
    total_count
  }
}
```

**Response:**

```json
{
  "data": {
    "dynamicBlocks": {
      "items": [
        {
          "uid": "MQ==",
          "content": {
            "html": "<h2><strong>SAVE 20%</strong></h2>\r\n<p>(some restrictions apply)</p>\r\n<p>&nbsp;</p>"
          }
        },
        {
          "uid": "Mg==",
          "content": {
            "html": "<p><img src=\"{{media url=&quot;wysiwyg/save20.png&quot;}}\" alt=\"save 20% red\"></p>"
          }
        }
      ],
      "page_info": {
        "current_page": 1,
        "page_size": 20,
        "total_pages": 1
      },
      "total_count": 2
    }
  }
}
```

**cmsPage query response:**

> The [`cmsPage` query]({{page.baseurl}}/graphql/queries/cms-page.html) would return the following code for a CMS page that contains the dynamic block in this example with the `uid` of `MQ==`. The response has been reformatted for readability.

Reviewer 3: ```suggestion
The [`cmsPage` query]({{page.baseurl}}/graphql/queries/cms-page.html) returns the following code for a CMS page that contains the dynamic block; in this example, a block with the `uid` of `MQ==`. The response has been reformatted for readability.
```

  data-types=\"\"
  data-display-mode=\"fixed\"
  data-ids=\"1\"
  data-rotate=\"\"
  data-store-id=\"1\"
  data-uids=\"MQ==\">\n

  <ul class=\"banner-items\"
    data-bind=\"afterRender: registerBanner\">\n
    
    <!-- ko foreach: getItems833e4819d6c46ab41e9910f17dc04f72329cb84f1b0dc3aa76d43bcb11d605a6() -->\n   
    
    <li class=\"banner-item\" 
      data-bind=\"attr: {'data-banner-id': $data.bannerId}\">\n
      
      <div class=\"banner-item-content\" data-bind=\"bindHtml: $data.html\"></div>\n
      
    </li>\n
    
    <!-- /ko -->\n
  </ul>\n
</div>
```

## Input attributes

Attribute | Data type | Description
--- | --- | ---
`dynamic_block_uids` | [ID] | An array of dynamic block UIDs to filter on
`locations` | [DynamicBlockLocationEnum] | An array indicating the locations the dynamic block can be placed. The possible values are CONTENT, HEADER, FOOTER, LEFT, and RIGHT. If this attribute is not specified, the query returns all locations
`type` | DynamicBlockTypeEnum! | A value indicating the type of dynamic block to filter on. The possible values are CART_PRICE_RULE_RELATED, CATALOG_PRICE_RULE_RELATED, and SPECIFIED

> ## Output attributes

Reviewer 3: Space between table and header.

> The `DynamicBlocks` output object contains the following attributes.

Reviewer 3: ```suggestion
The `DynamicBlocks` output object contains the following attributes:
```
Same with the following two.
